src/xb-arena.c

```diff
@@ -0,0 +1,137 @@
+
+#include "xb-arena.h"
+
+#include <stddef.h>
+#include <string.h>
+#include <sys/types.h>
+#include "glib.h"
+
+XbArena xb_arena_new(void) {
+    return (XbArena){
+        .n_chunks = 0,
+        .chunks_cap = 8,
+        .chunks = malloc(8 * sizeof(uint8_t*)),
+        .tail = NULL,
+        .current_chunk_end = NULL,
+    };
+}
+
+static void xb_arena_add_chunk(XbArena* arena, uint8_t* chunk) {
+    /* Grow chunklist if needed */
+    if (arena->n_chunks + 1 >= arena->chunks_cap) {
+        uint8_t** new_chunklist = malloc(arena->chunks_cap * sizeof(uint8_t*) * 2);
+        memcpy(new_chunklist, arena->chunks, arena->n_chunks * sizeof(uint8_t*));
+        free(arena->chunks);
+        arena->chunks = new_chunklist;
+        arena->chunks_cap *= 2;
+    }
+    arena->chunks[arena->n_chunks] = chunk;
+    arena->n_chunks++;
+}
+
+void* xb_arena_alloc(XbArena* arena, size_t len) {
+    if(len == 0) return NULL;
+    
+    /* Treat allocations larger than a chunk size as their own chunk
+     * It's not worth trying to optimize those. If this case hits a lot, try
+     * increasing the arena chunk size. */
+    if(len > XB_ARENA_CHUNKSIZE){
+        uint8_t* chunk = malloc(len);
+        xb_arena_add_chunk(arena, chunk);
+        /* No need to invalidate tail or current_chunk_end, there's still space
+        in the current chunk for more allocations */
+        return chunk;
+    }
+    /* Round allocations up to nearest whole machine word
+     * If we didn't do this, allocated structs may cost extra
+     * to access because of unaligned load/stores and may
+     * straddle multiple cache lines */
+    len = (len+sizeof(ssize_t)-1) & ~(sizeof(ssize_t)-1);
```

Comment:
If available, I think you should align to the alignment of [`max_align_t`](https://en.cppreference.com/w/c/types/max_align_t), which aiui should give you the same fundamental alignment guarantees as `malloc()` itself.

Reply:
However, you may want to consider doing the alignment at the _head_ of an allocation (i.e. setting `alloc = arena->tail + alignment_offset`) rather than by incrementing the `len` of the previous allocation, as this will allow for tighter packing of small allocations: the alignment requirement is limited by the requested `len` of an allocation, so small allocations don’t need to be 16-byte aligned (for example).